base/mpsc_intrusive_queue.h

```diff
@@ -91,9 +97,35 @@ template <typename T> std::pair<T*, bool> MPSCIntrusiveQueue<T>::PopWeak() noexc
   T* next = MPSC_intrusive_load_next(*head);  // load(std::memory_order_acquire)
   if (stub() == head) {
     if (nullptr == next) {
-      // empty
-      return {nullptr, true};
+      // Empty state.
+      // Caveat: if Push() called on an empty queue but has not crossed the blocking point yet,
+      // we may reach this condition because head_ is a stub and stub.next is nullptr.
+      // Unfortunately it may lead to a bizarre scenario where the arbitrary number of
+      // subsequent pushes will fully complete, but the queue will still be observerd
+      // as empty by the consumer because the chain will be cut by the Push that is stuck updating
+      // the stub.
+      //
+      // More comments: if we had a single Push that is not completed yet, then returning
+      // an empty state is fine. The problem arises when we have multiple pushes in parallel
+      // and the first one has not completed yet, but others did.
+      // Moreover, one of the pushes could even be linearized with this PopWeak call
+      // using locks/barriers, and still would not be be observed.
+      //
+      // To disambiguite, we load the tail_ and check if it is the same as the head.
+      // MO is not important because for operations that are externally linearized,
+      // the order is already guaranteed.
+      // To sum up:
+      // 1. if tail is not head, it is quaranteed that the queue is not empty.
+      // 2. If Push(Item) has finished, and was linearized with PopWeak somehow,
+      //    is guaranteed that we will observe the updated tail_.
+      // 3. Otherwise, it's just an optimization that may or may not return the updated tail.
+      //    Remember that standard guarantees that stores will eventually be visible with any MO,
+      //    therefore we may see here stale (head), or the updated tail even if Push has finished.
+      T* tail = tail_.load(std::memory_order_relaxed);
+      return {nullptr, tail == head};
```

Comment:
the memory order is not the issue here. The issue is that we want to avoid these situations for 100%, it's not possible with Queue api unless we linearize using some external conditions.
i.e. even with a single Push that is completed, it's not guaranteed that PopWeak returns non-empty, because 
store/load are not strong consistent withing themselves. I wanted to empasize this. Please suggest the better comment :)